### ISSUE TITLE
Octopus Roadmap

### ISSUE THREAD

**Issue by PaulStovell** [OctopusDeploy/Issues#1]
_Issue by **[PaulStovell](https://github.com/PaulStovell)** from Sunday Aug 18, 2013 at 02:40 GMT_
_Originally opened as https://github.com/OctopusDeploy/OctopusDeploy/issues/10_

---

For a long time I imagined our 2.0 release would be:

> Here's Octopus 2.0 with 84 new big features

Instead, let's focus on:

> Here's Octopus 2.0. It has some breaking changes, but it's a platform upon which we'll be able to build 84 new big features

For 2.0 I think customers will be understanding if the upgrade process is painful, but I think this is really our last chance to make any breaking, painful changes for a few years. 

To make this happen, we'll need look at all of the feature ideas we've currently got in mind for the near future, and get a rough idea of how they're going to fit in with the current architecture. If they won't be able to be added without serious breaking changes/pain for our users, let's make some of those changes now. We don't need to build the features in yet, but we'll need to have a platform capable of letting us build them in later releases. 

Of course for 2.0 we'll be building a migration tool so that Octopus 1.0 customers will be able to go to 2.0. But that migration tool is going to be painful to build, and I don't want to have to build a 2.1 migration tool, and a 2.2 migration tool. 

So let's pivot, and focus on getting a release out that will be our platform for the future. Thus far we've been developing Octopus 2.0 in a cone of silence. To prepare for a release, let's gradually open it up to people, starting with a private preview release for early adopters, then slowly moving to a feature complete release. 
## Milestone 1: Octopus 2.0 Preview

We'll email our mailing list asking for people who want to try an early preview. They'll get a link to our TeamCity server where they can download the automated builds. For these releases we're mainly looking for feedback on:
- Whether the application even runs
- Security/configuration issues arising from: 
  1. The decision to use Nancy as the UI stack
  2. the new communication stack
- Experiences with the new UI and API

At a minimum, we'll need:
- [ ] Working end to end deployments for the existing step types
- A UI capable of:
  - [ ] Defining environments and machines
  - [ ] Defining projects, steps, and variables
  - [ ] Creating and deploying releases
  - [ ] Setting up NuGet feeds
  - [ ] Viewing task logs
- [ ] A basic installer and configuration tool, enough to get the server and Tentacles running
- [ ] A secure transport solution for Octopus <-> Tentacle communication

Out of scope for this release will be:
- Multiple users/permissions/teams - assume there's just one Administrator user with all permissions. All actions will be locked down to admin-only
- Linux/mono support
- Any other new features, like multi-tenancy etc.
- Advanced install/configuration options
- Any form of migration tool

---
## Milestone 2: Octopus 2.0 Beta

Since Octopus 2.0 is our last chance to make breaking/painful changes, we'll need to make room for them now. Although we might not implement the features fully, we'll need to make sure we have a platform on which to build them that won't cause pain.
- Decide what other breaking/painful changes we need to make
- Ensure we have the ability to handle them in our platform, either now or at a later date
- Add multiple users, permissions and teams 
- Get the UI "feature complete"

Initially, this work will be visible to our preview users in the TeamCity builds. As we start to look more polished and feature compete, we'll widen the audience and make these releases downloadable from the OctopusDeploy.com website. 

---
## Milestone 3: Octopus 2.0

Before we can call Octopus 1.0 "done", we need a migration tool for 1.0 customers. This will be a seperate download, and it will:
- Allow users to select their 1.0 data folder
- Upgrade it to the 2.0 schema
- Automatically upgrade all Tentacles

The reason for putting it in a seperate tool will be to reduce the amount of legacy baggage we need to carry around in the 2.0 codebase. 

---
## Octopus 2.1 and beyond

With 2.0 out of the way, we can start to get into a rhythm of consistent releases with new features and no breaking changes. I'd love to see us doing a new release every second Tuesday with new features and fixes. For example:
- Octopus 2.1: Embedded NuGet server
- Octopus 2.2: Multi-tenancy
- Octopus 2.3: Linux/Mono support
